java/arcs/android/common/resurrection/ResurrectorService.kt

```diff
@@ -0,0 +1,204 @@
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.android.common.resurrection
+
+import android.app.Service
+import android.content.ComponentName
+import android.content.Intent
+import android.os.Bundle
+import androidx.annotation.VisibleForTesting
+import arcs.core.storage.StorageKey
+import arcs.core.util.guardWith
+import java.io.PrintWriter
+import kotlinx.coroutines.CoroutineName
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.async
+import kotlinx.coroutines.cancelChildren
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+/**
+ * Extension point for [Service]s which wish to be capable of resurrecting their clients.
+ */
+abstract class ResurrectorService : Service() {
+    /**
+     * The filename of the SQLite database used to persist [ResurrectionRequest]s.
+     */
+    protected open val resurrectionDatabaseName: String = DbHelper.RESURRECTION_DB_NAME
+
+    private val job = Job() + Dispatchers.IO + CoroutineName("ResurrectorService")
+    private val dbHelper: DbHelper by lazy { DbHelper(this, resurrectionDatabaseName) }
+
+    private val mutex = Mutex()
+    private var registeredRequests: Set<ResurrectionRequest>
+        by guardWith(mutex, setOf())
+    private var registeredRequestsByNotifiers: Map<StorageKey?, Set<ResurrectionRequest>>
+        by guardWith(mutex, mapOf())
+    @VisibleForTesting lateinit var loadJob: Job
+
+    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
+        intent?.action?.takeIf { it == ACTION_RESET_REGISTRATIONS }?.let {
+            loadJob = CoroutineScope(job).resetRequests()
+        } ?: ResurrectionRequest.createFromIntent(intent)?.let {
+            loadJob = CoroutineScope(job).registerRequest(it)
+        } ?: ResurrectionRequest.componentNameFromUnrequestIntent(intent)?.let {
+            loadJob = CoroutineScope(job).unregisterRequest(it)
+        } ?: {
+            loadJob = CoroutineScope(job).launch {
+                val needToLoad = mutex.withLock { registeredRequests.isEmpty() }
+                if (needToLoad) loadRequests()
+            }
+        }()
+
+        return super.onStartCommand(intent, flags, startId)
+    }
+
+    override fun onDestroy() {
+        super.onDestroy()
+        job.cancelChildren()
+    }
+
+    /**
+     * Makes [Context.startService] or [Context.startActivity] calls to all clients who are
+     * registered for the specified [events] (or are registered for *all* events).
+     */
+    @VisibleForTesting(otherwise = VisibleForTesting.PROTECTED)
+    suspend fun resurrectClients(vararg events: StorageKey) = resurrectClients(events.toList())
```

Comment:
nit - shouldn't the parameter name match the type?

Reply:
Good point, I'll tweak in a followup pr.